### PR TITLE
[Core] Add Back Pressure to GCS's gRPC Server

### PR DESCRIPTION
**Pull request by lixin-wei**
## Why are these changes needed?

Prior to this PR, we create a ServerCall in HandleRequestImpl() and delete it when it is replied and pulled again from the completion queue.

However, it will cost some time from handling the request to deleting it. If request TPS is high, tons of ServerCalls will be created during this time, which will cause a huge memory usage(in our case, up to 64+GB, caused OOM).

This PR fixes this issue, please have a look, thanks!

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by edoakes** [comment]
@clarkzinzow can you also take a look at this? I know you recently dove into the gRPC completion queue code.

**Comment by edoakes** [comment]
@lixin-wei looks like this failed to compile in the Travis build.

**Review by edoakes — COMMENTED** [review]
If I understand correctly, the behavior change here is that we now create the new `server_call` _after_ the reply callback has been called, where previously we were creating the new call _before_ even invoking the RPC handler. Please correct me if I'm wrong.

What will happen if we drain the queue of `server_call`s in this case (e.g., we have a buffer of 100 `server_call`s and we have 100 RPCs that are all outstanding)? I would assume that this would block the RPC from being handled until one of the outstanding RPCs finishes. This may cause problems in the case of e.g., asyncio actors where there may be thousands of outstanding actor calls concurrently. If that's the case, we will probably need to either (1) warn users that there's a max number of outstanding actor calls and probably default to something quite high or (2) make the limit configurable.

If we're going to merge this, we should definitely run the full suite of nightly/release benchmarks and stress tests on it. A good start would be to post the microbenchmark results before and after the PR.

**Comment by lixin-wei** [comment]
> I would assume that this would block the RPC from being handled until one of the outstanding RPCs finishes.

Yeah, that's exactly what 'back pressure' means.

> This may cause problems in the case of e.g., asyncio actors where there may be thousands of outstanding actor calls concurrently. 

In this case, user actors should put the requests into their own queues, and return immediately from GRPC's handling function. From an RPC framework's vision, back pressure is necessary.

**Comment by lixin-wei** [comment]
And I agree with make the limit configurable,  I've committed it.

**Comment by ericl** [comment]
@lixin-wei can you say more about what scenario triggers OOM on the raylet? AFAIK, all of Ray's components implement internal back-pressure, so it should not be possible to trigger such an overload because we have backpressure at higher levels.

I also share @edoakes 's concerns about this introducing distributed deadlocks. The unfortunate matter is that certain components require many active RPCs (e.g. async long-polling waiting for object locations). The number of async RPCs can scale with the number of workers in the cluster (e.g., up to tens of thousands).

What if we made this opt in--- each server/handler can request a max limit, and we can only enforce the limit on components that aren't handling async RPCs?

**Review by ericl — CHANGES_REQUESTED** [review]
[see comment]

**Comment by lixin-wei** [comment]
> can you say more about what scenario triggers OOM on the raylet? 
@ericl  It's in GCS, not raylet. We have a job with 50k+ actors whose creation task's argument is 100+ ActorHandles. In ActorHandle's deserialization, a GetActorInfo RPC will be sent to GCS. When this job is starting, GCS will face a 50k/s TPS of  GetActorInfo RPC. And memory usage hit 64GB+, which caused OOM.

**Comment by lixin-wei** [comment]
> can you say more about what scenario triggers OOM on the raylet? 

@ericl  It's in GCS, not raylet. We have a job with 50k+ actors whose creation task's argument is 100+ ActorHandles. In ActorHandle's deserialization, a GetActorInfo RPC will be sent to GCS. When this job is starting, GCS will face a 50k/s TPS of  GetActorInfo RPC. And memory usage hit 64GB+, which caused OOM.

**Comment by lixin-wei** [comment]
> What if we made this opt in--- each server/handler can request a max limit, and we can only enforce the limit on components that aren't handling async RPCs?

I think it's a good idea. Actually, I just want to fix GCS's OOM, we can apply this to GCS first, leave other components as they are.

**Comment by ericl** [comment]
I see, it makes sense to enable this by default for everything in GCS-- AFAIK, there are no long-running async calls to the GCS.

**Comment by rkooo567** [comment]
Actually, we have 1 RPC now PollActorOutOfScope (just lyk). And if we replace Redis pubsub, that will also the async long polling (though gcs server will be the receiver)

**Comment by lixin-wei** [comment]
@rkooo567 I can't find PollActorOutOfScope in our code base, what is it?  Could you say more about it?

**Comment by lixin-wei** [comment]
@ericl @edoakes @rkooo567  So do you agree with it if I make this configurable and only apply it to GCS? If you agree I'll proceed with my PR.

**Comment by rkooo567** [comment]
It's https://github.com/ray-project/ray/blob/58c4fe727c56ff7aff182db738e629fe1fe6d4b7/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L509

This means we cannot apply this to gcs server right? unless we can remove this RPC

**Comment by lixin-wei** [comment]
@rkooo567 If I understand correctly, this is an RPC from GCS to CoreWorker? Adding back pressure to GCS's gRPC server won't affect this.

>if we replace Redis pubsub, that will also the async long polling (though gcs server will be the receiver)

what did you mean with 'replace Redis pubsub'? Would you please say more about this? I don't know much about the context.

**Comment by rkooo567** [comment]
Ah, I see. it is only for the server. When I said replace Redis, I meant this PR; https://github.com/ray-project/ray/pull/16723 (this will use long polling from other components -> GCS for pubsub instead of Redis). I think this will be done in a couple month, so if you can make backpressure configurable, it should be fine (and default could be no backpressure for OSS).

But I also wonder if there's way to flagging each call somehow from the client side? If it is possible, we can probably only mark long-polling RPC to be not included for the backpressure (and we technically shouldn't have many long polling RPC because it causes memory leak if there are many)

**Comment by ericl** [comment]
Applying to GCS sounds good to me!

**Comment by lixin-wei** [comment]
@rkooo567 Yeah, If we replace Redis with gRPC based pubsub, there will be a lot of long pollings to GCS.

> But I also wonder if there's way to flagging each call somehow from the client side? 

It's not easy to do this. But it's easy to make it configurable for every service(disabled by default). I'll add back pressure to all GCS's services now. @wuisawesome you don't need to worry about this in <https://github.com/ray-project/ray/pull/16723> because it's disabled by default, which won't affect the service you created in that PR.

**Comment by rkooo567** [comment]
SGTM!

**Comment by lixin-wei** [comment]
I'm done. Would please have a look?

(BTW, the code of RPC part is pretty nasty, I've a hard time reading it. We should refactor it in the future.

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -460,3 +460,6 @@ RAY_CONFIG(uint64_t, resource_broadcast_batch_size, 512);
 // If enabled and worker stated in container, the container will add
 // resource limit.
 RAY_CONFIG(bool, worker_resource_limits_enabled, false)
+
+/// ServerCall instance number of each RPC service
+RAY_CONFIG(int64_t, gcs_default_back_pressure_limit_per_service, 100)
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:465]:
```suggestion
RAY_CONFIG(int64_t, gcs_max_active_rpcs_per_service, 100)
```

File: src/ray/rpc/grpc_server.h [hunk 2]
```diff
@@ -36,6 +36,15 @@ namespace rpc {
           #SERVICE ".grpc_server." #HANDLER));                                  \
   server_call_factories->emplace_back(std::move(HANDLER##_call_factory));
 
+#define RPC_SERVICE_HANDLER_WITH_BACK_PRESSURE(SERVICE, HANDLER, BACK_PRESSURE_NUM) \
```
Comment (ericl) [inline src/ray/rpc/grpc_server.h:39]: Consider consolidating this with the original, and explicitly passing a limit of -1 for services that don't support backpressure.

File: src/ray/rpc/server_call.h [hunk 3]
```diff
@@ -95,6 +97,9 @@ class ServerCallFactory {
   /// corresponding type of requests.
   virtual void CreateCall() const = 0;
 
+  /// Get the maximum request number to handle at the same time. -1 means no limit.
+  virtual int64_t GetBackPressureLimit() const = 0;
```
Comment (ericl) [inline src/ray/rpc/server_call.h:101]:
```suggestion
  virtual int64_t GetMaxActiveRPCs() const = 0;
```

File: src/ray/rpc/server_call.h [hunk 4]
```diff
@@ -155,10 +160,13 @@ class ServerCallImpl : public ServerCall {
     // NOTE(hchen): This `factory` local variable is needed. Because `SendReply` runs in
     // a different thread, and will cause `this` to be deleted.
     const auto &factory = factory_;
-    // Create a new `ServerCall` to accept the next incoming request.
-    // We create this before handling the request so that the it can be populated by
-    // the completion queue in the background if a new request comes in.
-    factory.CreateCall();
+    if (factory.GetBackPressureLimit() == -1) {
```
Comment (ericl) [inline src/ray/rpc/server_call.h:163]:
```suggestion
    if (factory.GetMaxActiveRPCs() == -1) {
```

File: src/ray/rpc/server_call.h [hunk 5]
```diff
@@ -267,20 +277,24 @@ class ServerCallFactoryImpl : public ServerCallFactory {
...
+  /// means no limit.
   ServerCallFactoryImpl(
       AsyncService &service,
       RequestCallFunction<GrpcService, Request, Reply> request_call_function,
       ServiceHandler &service_handler,
       HandleRequestFunction<ServiceHandler, Request, Reply> handle_request_function,
       const std::unique_ptr<grpc::ServerCompletionQueue> &cq,
-      instrumented_io_context &io_service, std::string call_name)
+      instrumented_io_context &io_service, std::string call_name,
+      int64_t back_pressure_limit = -1)
```
Comment (ericl) [inline src/ray/rpc/server_call.h:289]:
```suggestion
      int64_t max_active_rpcs = -1)
```

File: src/ray/rpc/server_call.h [hunk 6]
```diff
@@ -315,6 +331,10 @@ class ServerCallFactoryImpl : public ServerCallFactory {
 
   /// Human-readable name for this RPC call.
   std::string call_name_;
+
+  /// Maximum request number to handle at the same time.
+  /// -1 means no limit.
+  uint64_t back_pressure_limit_;
```
Comment (ericl) [inline src/ray/rpc/server_call.h:337]:
```suggestion
  uint64_t max_active_rpcs_;
```

**Review by ericl — CHANGES_REQUESTED** [review]
LGTM, only stylistic requests

**Comment by rkooo567** [comment]
The failure from `tests:test_placement_group ` seems quite suspicious.

**Comment by lixin-wei** [comment]
@rkooo567 Yeah, I'm trying to find out why.

**Comment by lixin-wei** [comment]
I can't reproduce the failed test in my local machine, pull master and retry.

**Comment by lixin-wei** [comment]
CI passed, would you please have a look? The streaming test failure is a known issue related to the CI framework.

You may take attention to this change:

https://github.com/ray-project/ray/pull/17427/files#diff-116723aef6c907367b1b8591aa9e2ad395e7e6a2e9a94c27ec16633c877236b7R180-R184

I add no limit to Register/Create actor handler because CreateActor is a long polling RPC. It will be replied to until raylet successfully schedules the creation task. If there are no resources, GCS will hold a lot of CreateActor RPCs, once it reaches the backpressure limit, GCS won't receive any CreateActor requests anymore. But the incoming actors might be schedulable(if the actor requires small resources).

**Review by ericl — APPROVED** [review]

**Pull request merged**